swirl_lm/boundary_condition/immersed_boundary_method.py

```diff
@@ -317,6 +328,507 @@ def _check_mesh_size(a, b):
+        temp_mask.append(
+            common_ops.gather(tf.transpose(ib_interior_mask, perm=[1, 2, 0]),
+                tf.stack([lower_left_index[0] + i,
+                          lower_left_index[1] + j,
+                          lower_left_index[2] + k],
+                axis=1))
+        )
+
+      fluid_corner_point_mask = tf.nest.map_structure(
+          lambda _v: tf.where(_v < 0.75, tf.zeros_like(_v), tf.ones_like(_v)),
+          temp_mask,
+      )
+
+      # We mask out all enclosing grid nodes that lay in the solid domain by
+      # multiplying their distance to the image point with zero, while all grid
+      # nodes in the solid are masked with 1 (i.e., our
+      # `fluid_corner_point_mask`).
+      node_dist = tf.nest.map_structure(tf.math.multiply, node_dist,
+                                        fluid_corner_point_mask)
+
+      # Flag all image points that have less than 8 enclosing grid nodes
+      # in the FLUID domain. If all enclosing grid nodes for a given image point
+      # are inside the fluid domain, the sum of `fluid_corner_point_mask` will be
+      # exactly 8 (since there are 8 enclosing nodes in 3D space for an image
+      # point). However, if one or more grid nodes are inside the solid domain,
+      # the corresponding sum of `fluid_corner_point_mask` will be less than 8,
+      # which indicates that we have a degenerated cell. Note that
+      # `fluid_corner_point_mask` already flags every node with `1` if inside the
+      # fluid domain, and `0` if inside the solid domain.
+      # Source: Iaccarino and Verzicco, Applied Mechanics Reviews 56(3), 2003.
+      degenerated_cell_mask = tf.nest.map_structure(
+          lambda v: tf.where(v < 7.5, tf.ones_like(v), tf.zeros_like(v)),
+          tf.math.reduce_max(fluid_corner_point_mask, axis=0)
+      )
+
+      # Add the boundary interecpt (BI) as an additional point, if the enclosing
+      # cell of a given image point is degenerated (i.e. has less than 8
+      # enclosing nodes in the fluid domain).
+      # Essentially, we are adding the distance between the ghost point(!) and
+      # the IB to `node_dist` which will now have 9 rows (8 distances between the
+      # image point and its surrounding grid nodes, and 1 additional row for the
+      # distance between the image point and the IB). However, the 9th row will
+      # only have non-zero distances, if the enclosing grid cell is degenerated,
+      # otherwise, the distance will be zero. As we will always use all 9 rows of
+      # `node_dist` to compute the interpolated value on the image point, we want
+      # to make sure that the BI is excluded for an eclosing cell that has all 8
+      # grid nodes in the fluid domain, and included otherwise.
+      node_dist.append(
+          tf.nest.map_structure(
+            lambda d, m: tf.math.pow(
+                tf.math.divide(1.0, d + _EPSILON), _EXP) * m,
+            gp_dist,
```

Comment:
Let's make a note here that gp_dist is used because it's the same as the ip_dist.